src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,424 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+              in = bs->step_over_gc_barrier(in);
+              if (in->Opcode() == Op_DecodeN) {
+                in = in->in(1);
+              }
+              assert(in->Opcode() == Op_LoadP || in->Opcode() == Op_LoadN, "");
+              assert(C->get_alias_index(in->adr_type()) == C->get_alias_index(TypeAryPtr::OOPS), "");
+              Node* addp1 = in->in(MemNode::Address);
+              assert(addp1->is_AddP(), "");
+              assert(addp1->in(AddPNode::Base)->uncast()->is_Proj() &&
+                     addp1->in(AddPNode::Base)->uncast()->in(0)->as_CallJava()->method()->intrinsic_id() ==
+                     vmIntrinsics::_scopedValueCache, "");
+              if (scoped_value_cache == nullptr) {
+                scoped_value_cache = addp1->in(AddPNode::Base)->uncast()->in(0)->as_Call();
+              } else {
+                assert(scoped_value_cache == addp1->in(AddPNode::Base)->uncast()->in(0), "");
+              }
+              assert(in->in(MemNode::Memory)->is_Proj() && in->in(MemNode::Memory)->in(0) == scoped_value_cache, "");
+              Node* addp2 = addp1->in(AddPNode::Address);
+              Node* offset1 = addp1->in(AddPNode::Offset);
+              intptr_t const_offset = offset1->find_intptr_t_con(-1);
+              BasicType bt = TypeAryPtr::OOPS->array_element_basic_type();
+              int shift = exact_log2(type2aelembytes(bt));
+              int header = arrayOopDesc::base_offset_in_bytes(bt);
+              assert(const_offset >= header, "");
+              const_offset -= header;
+
+              Node* index = kit.gvn().intcon(const_offset >> shift);
+              if (addp2->is_AddP()) {
+                assert(!addp2->in(AddPNode::Address)->is_AddP() &&
+                       addp2->in(AddPNode::Base) == addp1->in(AddPNode::Base),
+                       "");
+                Node* offset2 = addp2->in(AddPNode::Offset);
+                assert(offset2->Opcode() == Op_LShiftX && offset2->in(2)->find_int_con(-1) == shift, "");
+                offset2 = offset2->in(1);
+#ifdef _LP64
+                assert(offset2->Opcode() == Op_ConvI2L, "");
+                offset2 = offset2->in(1);
+                if (offset2->Opcode() == Op_CastII && offset2->in(0)->is_Proj() &&
+                    offset2->in(0)->in(0) == get_cache_iff) {
+                  ShouldNotReachHere();
+                  offset2 = offset2->in(1);
+                }
+#endif
+                index = kit.gvn().transform(new AddINode(offset2, index));
+              }
+
+              if (get_first_iff == nullptr) {
+                get_first_iff = c->as_If();
+                first_index = index;
+              } else {
+                assert(get_second_iff == nullptr, "");
+                get_second_iff = c->as_If();
+                second_index = index;
+              }
+            }
+          } else if (c->is_RangeCheck()) {
+            // Kill the range checks as they are known to always succeed
```

Comment:
> And what is the argument for why they always succeed?

Because of the way the cache is constructed:
```
              theCache = new Object[CACHE_TABLE_SIZE * 2];
              setScopedValueCache(theCache);
 ```
and
```
            CACHE_TABLE_SIZE = cacheSize;
            SLOT_MASK = cacheSize - 1;
```
accessed with:
```
            int n = (hash & Cache.SLOT_MASK) * 2;
            if (objects[n] == this) {
                return (T)objects[n + 1];
            }
            n = ((hash >>> Cache.INDEX_BITS) & Cache.SLOT_MASK) * 2;
            if (objects[n] == this) {
                return (T)objects[n + 1];
            }
```
> How do we know we do not accidentally kill a unrelated RangeCheck?

They are the only RangeChecks in `ScopedValue.get()`. Of course, because the c2 code pattern matches the IR of `ScopedValue.get()` it assumes it has a certain shape. The java code and c2 code have to stay in sync.